### PR TITLE
Check type at setting Variable.grad

### PR DESCRIPTION
**Pull request by beam2d**
It fixes the second point of #209. `Variable.grad` is now a property whose setter checks the type, dtype, and shape of the argument. This PR also contains several fixes on Function implementations and some tests.

### CODE REVIEW

File: chainer/variable.py [hunk 1]
```diff
@@ -90,6 +90,24 @@ def label(self):
...
 
+    @property
+    def grad(self):
+        return self._grad
+
+    @grad.setter
+    def grad(self, g):
+        if g is not None:
+            if not isinstance(g, type(self.data)):
+                raise TypeError('Type of data and grad mismatch: %s != %s'
```
Comment (unnonouno) [inline chainer/variable.py]: How do users fix this problem?
It's a bug?

File: tests/test_variable.py [hunk 2]
```diff
@@ -137,5 +140,30 @@ def test_invalid_value_type(self):
         with self.assertRaises(AssertionError):
             chainer.Variable(1)
 
+    def test_grad_type_check_pass(self):
+        a = chainer.Variable(np.empty((3,), dtype=np.float32))
+        a.grad = np.ndarray((3,), dtype=np.float32)
+
+    def test_grad_type_check_type(self):
```
Comment (unnonouno) [inline tests/test_variable.py]: How about making a test for `dtype` and a test for `shape`?

**Comment by unnonouno** [comment]
I wrote some comments.

File: chainer/variable.py (hunk 1 above)
Comment (beam2d) [inline chainer/variable.py]: It depends on the case. If the user manually sets the grad and its type is wrong, then the user should fix this problem. If there is a function returns a grad with the wrong type, it's a bug. The user may distinguish these situations by watching the stack trace, though is a bit complicated task. I have no idea to distinguish them at this line.

File: tests/test_variable.py (hunk 2 above)
Comment (beam2d) [inline tests/test_variable.py]: `test_grad_type_check_dtype` and `test_grad_type_check_shape` are the ones. Aren't they enough?

**Comment by beam2d** [comment]
I updated the error messages of the grad setter.

File: tests/test_variable.py (hunk 2 above)
Comment (unnonouno) [inline tests/test_variable.py]: Ok, no problem.

**Comment by unnonouno** [comment]
It looks good to me. How do you think about broad casting problem?

File: tests/test_variable.py [hunk 3]
```diff
@@ -137,5 +140,30 @@ def test_invalid_value_type(self):
         with self.assertRaises(AssertionError):
             chainer.Variable(1)
 
+    def test_grad_type_check_pass(self):
+        a = chainer.Variable(np.empty((3,), dtype=np.float32))
+        a.grad = np.ndarray((3,), dtype=np.float32)
```
Comment (delta2323) [inline tests/test_variable.py]: Is there any tool that is equivalent to `assert_nothing_raised` in Ruby.
I am wondering if this test case can handle the exception when this assignment raises it.

Comment (beam2d) [inline tests/test_variable.py]: AFAIK, there is no such function that checks if the code raises nothing. When the code raises an exception without catch, the test is failed with the stack trace as same as usual assertion failures.

**Comment by beam2d** [comment]
@unnonouno You mean broadcasting between `data` and `grad` of the same variable? That sounds an interesting idea. The `grad` is purely a matter of backward computation where the `data` array is fixed beforehand, so there is only one case: broadcasting a size-one axis of `grad`. It should only occur when the user manually sets the gradient. We can support such broadcasting, though I think the merit is limited compared to the effort to support it (note that PyCUDA does not support broadcasting). Therefore, I think we do not need to support broadcasting now.

**Comment by beam2d** [comment]
I merged the current master.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**